File: mp1.py
# source: https://github.com/KrulYuno/cmsc141_machine_problems/blob/master/mp1.py

"""
    Test Cases: Amount of times creating new two sets.
"""


class InputFileHandler:
    def __init__(self, filename):
        self._filename = filename
        self._lines = []

    def start_reading(self):
        with open(self._filename, "r") as file:
            for line in file.readlines():
                self._lines.append(line)

    def _print(self):
        for line in self._lines:
            print(line)
    
    def _get_line(self, line:int):
        return self._lines[line-1] if abs(line-1) < len(self._lines) else None


class _Set:
    def _list(self) -> set:
        return self._set_list

    def _print(self):
        for item in self._set_list:
            print(item)

    def _get_type(self):
        return self._set_type

    # def _init_type(self, value):
    #     types = {
    #         1: int,
    #         2: float,
    #         3: str,
    #         4: str,
    #         5: set,
    #         6: object
    #     }
    #     return types[value]

    def insert(self, value):
        if not isinstance(value, self._set_type):
            raise TypeError(f"set item is of only type {self._set_type}")
        self._set_list.add(value)
    
    def remove(self, value):
        pass

    def subset(self):
        pass
    
    def union(self):
        pass

    def intersection(self):
        pass

    def difference(self):
        pass

    def power_set(self):
        pass


class _SetTypeCreator:
    """
        This function returns the type of set: (int, double, char, string, set, object)
    """
    def init_set(value: int, value2: int) -> _Set:
        stypes = {
                        1: _SetInteger,
                        2: _SetDouble,
                        3: _SetChar,
                        4: _SetString,
                        5: _SetSets,
                        6: _SetObject                        
                        }
        return stypes[value]()

class _SetInteger(_Set):
    def __init__(self):
        self._set_type = int
        self._set_list = set()

class _SetDouble(_Set):
    def __init__(self):
        self._set_type = float
        self._set_list = set()

class _SetChar(_Set):
    def __init__(self):
        self._set_type = str
        self._set_list = set()
    
    def insert(self, value: str):
        if not isinstance(value, self._set_type):
            raise TypeError(f"set item is of only type {self._set_type}")
        self._set_list.add(value[0])

class _SetString(_Set):
    def __init__(self):
        self._set_type = str
        self._set_list = set()

class _SetSets(_Set):
    def __init__(self):
        self._set_type = set
        self._set_list = set()
    

    def insert(self, set_value: set):
        if not isinstance(set_value, self._set_type):
            raise TypeError(f"set item is of only type {self._set_type}")
        self._set_list.update(set_value)


class _SetObject(_Set):
    def __init__(self):
        self._set_type = object
        self._set_list = set()

# ----------------------------------------------------------------- #
class Interface:
    """
        This class should be handling how we navigate through the sets. By creating, modifying,
        removing items from a set. An API perhaps?

        Flow:
            1) Initialize the sets - test cases
            2) Choose the set type for the each test case - execute _SetTypeCreator($type$)
                -> must strictly only use the specified set type for the created set
            3) Fill the two sets with data (based on the $type$ of set you specified.)
    """

    def __init__(self, tests: int, file):
        """
            __init__ is step 0 or the initialization
            We load the inititials:
                test_cases      : Amount of test cases
                test_file       : file input of test cases
                set_case        : ?
        """
        self._test_cases = tests
        self._test_file = InputFileHandler(file)
        self._test_file.start_reading()



    def step_1(self):
        """
            Get the type of set for the elements:
                1               : int
                2               : double
                3               : char
                4               : string
                5               : set
                6               : object
        """
        value = list(map(str, input().split()))
        value.append(None)
        value[0] = int(value[0])

        self._set1 = _SetTypeCreator.init_set(value[0], value[1])
        self._set2 = _SetTypeCreator.init_set(value[0], value[1])

        

    def step_2(self):
        for item in list(map(self._set1._get_type(), input("set_items1: ").split())):
            self._set1.insert(item)
        for item in list(map(self._set2._get_type(), input("set_items2: ").split())):
            self._set2.insert(item)

        

        print(self._set1._list())
        print(self._set2._list())
        

def main():
    new_test = Interface(input("test_cases: "), "mpa1.in")
    new_test.step_1()
    new_test.step_2()
    
    


if __name__ == '__main__':
    main()

File: mp1.cpp
// source: https://github.com/KrulYuno/cmsc141_machine_problems/blob/master/mp1.cpp

/*
    Test Cases: Amount of times creating new two sets.
    Operators: insert, remove, subset, union, intersection, difference, power set

*/

#include <stdio.h>
#include <iostream>
#include <string>
#include <fstream>
#include <vector>
#include <set>
#include <iterator> // i use this for iterating sets ¯\_(ツ)_/¯
#include <bits/stdc++.h> // oh i need this for std::istringstream

class InputFileHandler{
private:
    std::vector<std::string> _list;
    std::string _filename;

public:
    InputFileHandler(std::string filename){
        this->_filename = filename;
    }

    void start_reading(){
        std::ifstream File(this->_filename);
        std::string str;
        
        while(std::getline(File, str)){
            this->_list.push_back(str);
        }
        File.close();
    }
    
    void _print(){
        for(int i=0; i<this->_list.size(); i++) 
            std::cout << this->_list[i] << std::endl;
    }

    std::string _get_line(int line){
        // return this->_list.at(line-1);   // we don't start at line 0, we start at line 1
        return this->_list.at(line);        // due to offset, we don't need to subtract 1 anymore
    }
    
};

template <class SetType>
class Set{
private:
    std::set<SetType> _set_list;

public:
    void _print(){
        for(auto i = this->_set_list.begin(); i != this->_set_list.end(); i++)
            std::cout << *i << std::endl;
    }

    std::set<SetType> _get_type(){
        return this->_set_list;
    }

    void insert(SetType value){
        this->_set_list.insert(value);
    }

    void remove(SetType value){
        this->_set_list.erase(value);
    }

    void subset(){

    }

    void s_union(){


    }

    void intersection(){

    }

    void difference(){


    }

    void power_set(){


    }
};

class Interface{
/*
    This class should be handling how we navigate through the sets. By creating, modifying,
    removing items from a set. An API perhaps?

    Flow:
        0) Initialize the sets like test cases and offset for each test cases
        1) Get the set type
        2) We fill in the two sets with X amount of data
        3) we get the X number of operations and do X operation
        4) Repeat 1
    
    Check the function calculate_offset() for more info

*/
private:
    std::string _filename;
    int _test_cases;
    int _type;
    std::vector<int> _offset;           // we can't tell which line the next case starts so we take its offset for every case
    int _case_id;

public:
    int calculate_offset(){
        // TODO: offset generator for test cases
        return 0;
    }

    Interface(std::string file, int c_id = 0){
        /*
            This is step_0 and step_1, we initiate the test cases and count the offset for each test cases.
        */
        this->_case_id = c_id;
        this->_filename = file;
        InputFileHandler File(this->_filename);
        File.start_reading();


        /*
            This code below stores the set type based on the id number:
            1 : int
            2 : double
            3 : char
            4 : string
            5 : set         ( this a set type ex: [5 1] or [5 3])
            6 : object

            the id number is not used for the sets but rather to keep track of the input
            for type 5, an x amount of recursion might occur if the input is [5 5]
        */

        this->_test_cases = std::stoi(File._get_line( 0 + this->_offset.at(this->_case_id) ));
        if (File._get_line(2).length() == 1)
            this->_type = std::stoi(File._get_line(  1 + this->_offset.at(this->_case_id) ));
        else {
            // i remember now, it's splitting the two numbers
            std::istringstream sub_str(File._get_line(  1 + this->_offset.at(this->_case_id) ));
            std::vector<std::string> v_str;

            std::string temp;
            while (sub_str >> temp)
                v_str.push_back(temp); // this should now have values like [5, 1]

            this->_type = std::stoi(v_str.at(0)); // copy the set type
        }

        // okay we proceed to step 2, fill the set with data...
        this->step_2();
     }

    void step_2(){
        // factory method.... design patterns... meh, let's just use switch case
        switch (this->_type){
        case 1:
            this->step_2_int_type();                
            break;
        
        case 2:
            this->step_2_double_type();                
            break;
        
        case 3:
            this->step_2_char_type();                
            break;
        
        case 4:
            this->step_2_string_type();                
            break;
        
        case 5:
            this->step_2_set_type();                
            break;
        
        case 6:
            this->step_2_objects_type();                
            break;
        
        default:
            break;
        }

    }

    void step_2_int_type(){

    }

    void step_2_double_type(){

    }

    void step_2_char_type(){

    }

    void step_2_string_type(){

    }

    void step_2_set_type(){

    }

    void step_2_objects_type(){
        
    }
};

int main(){
    Interface Han("mp1_test.in");

    return 0;
}